### PR TITLE
[IMP] point_of_sale, pos_self_order: remove replaceDataByKey method

### PR DESCRIPTION
**Pull request by davidmonnom**
Before this commit when we would like to replace a local network with the server one, we was using a specific method that was using a specific record key. That was because the local record created in the front end doesn't have any server id. So we was using an UUID.

Now, the classic load data method is able to replace the local record by the server one using the specific local key. So we don't need anymore the replaceDataByKey method.

taskId: 4167707

### CODE REVIEW

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 1]
```diff
@@ -917,12 +883,17 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {},
 
                 baseData[model][record.id] = record;
 
-                if (records[model][record.id]) {
-                    oldStates[model][record.id] = records[model][record.id].serializeState();
+                const oldRecords = indexedRecords[model][modelKey][record[modelKey]];
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:886]:
```suggestion
                const oldRecord = indexedRecords[model][modelKey][record[modelKey]];
```

File: addons/point_of_sale/static/src/app/models/data_service.js [hunk 2]
```diff
@@ -298,6 +297,17 @@ export class PosData extends Reactive {
...
+                // Ensure loaded orders are linked to the current session,
+                // this modification is done here because is the most low
+                // level method that is called when loading orders
+                for (const order of result["pos.order"]) {
+                    if (order.state === "draft") {
+                        order.session_id = odoo.pos_session_id;
+                    }
+                }
+            }
+
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/data_service.js:310]: This is very specific to "pos.order" and it should be placed somewhere else. `execute` is supposed to be agnostic.

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.js [hunk 3]
```diff
@@ -46,6 +46,16 @@ export class PaymentScreen extends Component {
...
         onMounted(this.onMounted);
+        onWillStart(() => {
+            const order = this.pos.get_order();
+
+            for (const payment of order.payment_ids) {
+                const pmid = payment.payment_method_id.id;
+                if (!this.pos.config.payment_method_ids.map((pm) => pm.id).includes(pmid)) {
+                    payment.delete({ backend: true });
+                }
+            }
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.js:57]: I think you can just put this logic inside "this.onMounted".

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 4]
```diff
@@ -294,13 +302,15 @@ export class Base {
...
+    const indexes = opts.databaseIndex || {};
+    const database = opts.databaseTable || {};
     const [inverseMap, processedModelDefs] = processModelDefs(modelDefs);
-    const records = mapObj(processedModelDefs, () => reactive({}));
-    const orderedRecords = mapObj(processedModelDefs, () => reactive([]));
+    const records = mapObj(processedModelDefs, () => reactive(new Map()));
     const callbacks = mapObj(processedModelDefs, () => []);
     const commands = mapObj(processedModelDefs, () => ({
         delete: new Map(),
+        unlink: new Map(),
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:313]: what's the difference between unlink and delete?

Comment (davidmonnom) [reply addons/point_of_sale/static/src/app/models/related_models.js:313]: Unlink will just unlink the record in the backend and delete will delete it

File: addons/pos_restaurant/static/src/overrides/models/pos_store.js [hunk 5]
```diff
@@ -44,59 +44,13 @@ patch(PosStore.prototype, {
...
-
-            ["config_id", "in", [...this.config.raw.trusted_config_ids, this.config.id]],
-            ["table_id", "in", data["table_ids"]],
-            ["session_id", "=", odoo.pos_session_id],
-        ]);
-        for (const table of data["order_count"]) {
-            tableByIds[table.id].uiState.changeCount = table.changes;
-            tableByIds[table.id].uiState.orderCount = table.orders;
-            tableByIds[table.id].uiState.skipCount = table.skip_changes;
-        }
```
Comment (caburj) [inline addons/pos_restaurant/static/src/overrides/models/pos_store.js:99]: This is from a recent bug fix and you are removing it. Did you test this change?

File: addons/point_of_sale/static/src/app/models/data_service.js [hunk 6]
```diff
@@ -28,7 +28,7 @@ export class PosData extends Reactive {
         this.custom = {};
         this.syncInProgress = false;
         this.mutex = markRaw(new Mutex());
-        this.records = {};
+        this.records = new Map();
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/data_service.js:31]: we don't need this change.

File: addons/point_of_sale/static/src/app/models/related_models.js [hunk 7]
```diff
@@ -917,11 +899,15 @@ export function createRelatedModels(modelDefs, modelClasses = {}, indexes = {},
...
-                    oldStates[model][record.id] = records[model][record.id].serializeState();
+                const oldRecord = indexedRecords[model][modelKey][record[modelKey]];
+                if (oldRecord) {
+                    oldStates[model][oldRecord[modelKey]] = oldRecord.serializeState();
                 }
 
                 const result = create(model, record, true, false, true);
+                if (oldRecord && oldRecord.id !== result.id) {
+                    oldRecord.delete();
+                }
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/related_models.js:910]: can't we use "update" here?

File: addons/pos_restaurant/models/pos_restaurant.py [hunk 8]
```diff
@@ -71,6 +71,15 @@ def sync_from_ui(self, name, background_color, config_id):
...
 
+    @api.model
+    def update_floor(self, floor_ids, config_id):
+        config = self.env['pos.config'].browse(config_id)
+        floors = self.browse(floor_ids)
+        return {
+            'restaurant.floor': floors.read(self._load_pos_data_fields(config.id)),
+            'restaurant.table': floors.table_ids.read(floors.table_ids._load_pos_data_fields(config.id)),
+        }
+
```
Comment (caburj) [inline addons/pos_restaurant/models/pos_restaurant.py:82]: seems unused

**Review by caburj — APPROVED** [review]

**Comment by adgu-odoo** [comment]
robodoo r+

**Pull request closed without merging**